clients/src/main/java/org/apache/kafka/common/protocol/types/ArrayOf.java

```diff
@@ -72,10 +74,10 @@ public Object read(ByteBuffer buffer) {
         else if (size < 0)
             throw new SchemaException("Array size " + size + " cannot be negative");
 
-        Object[] objs = new Object[size];
+        List<Object> objs = new ArrayList<>();
         for (int i = 0; i < size; i++)
-            objs[i] = type.read(buffer);
-        return objs;
+            objs.add(type.read(buffer));
+        return objs.toArray();
```

Comment:
> Checking for OOM is a security issue, so we cannot do that as the main approach. We need to do pre-validation.

@ijuma 
Then how about using a list instead of an array? I know this might have a negative impact on performance but it seems to resolve OOM issue.